File: frontend/src/utils/validators.ts
// Validação de CPF/CNPJ (dígitos verificadores) — espelha core/validators.py.

const digits = (v: string) => (v || '').replace(/\D/g, '')

export function validaCPF(cpf: string): boolean {
  const c = digits(cpf)
  if (c.length !== 11 || /^(\d)\1{10}$/.test(c)) return false
  for (const tam of [9, 10]) {
    let soma = 0
    for (let i = 0; i < tam; i++) soma += parseInt(c[i], 10) * (tam + 1 - i)
    let dig = (soma * 10) % 11
    if (dig === 10) dig = 0
    if (dig !== parseInt(c[tam], 10)) return false
  }
  return true
}

export function validaCNPJ(cnpj: string): boolean {
  const c = digits(cnpj)
  if (c.length !== 14 || /^(\d)\1{13}$/.test(c)) return false
  const tabelas: [number[], number][] = [
    [[5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2], 12],
    [[6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2], 13],
  ]
  for (const [pesos, pos] of tabelas) {
    let soma = 0
    for (let i = 0; i < pos; i++) soma += parseInt(c[i], 10) * pesos[i]
    const resto = soma % 11
    const dig = resto < 2 ? 0 : 11 - resto
    if (dig !== parseInt(c[pos], 10)) return false
  }
  return true
}

/** Aceita CPF (11 díg.) ou CNPJ (14 díg.); valida os dígitos verificadores. */
export function validaCpfCnpj(numero: string): boolean {
  const n = digits(numero)
  if (n.length === 11) return validaCPF(n)
  if (n.length === 14) return validaCNPJ(n)
  return false
}

File: core/validators.py
"""Validação de CPF/CNPJ (dígitos verificadores) — fonte única do sistema.

Usado no cadastro de oficina (CNPJ), no cadastro de cliente (CPF ou CNPJ) e
no gateway de pagamento (que rejeita documento inválido no PagBank)."""
import re


def so_digitos(valor: str) -> str:
    return re.sub(r'\D', '', valor or '')


def valida_cpf(cpf: str) -> bool:
    cpf = so_digitos(cpf)
    if len(cpf) != 11 or cpf == cpf[0] * 11:
        return False
    for tam in (9, 10):
        soma = sum(int(cpf[i]) * ((tam + 1) - i) for i in range(tam))
        dig = (soma * 10) % 11
        dig = 0 if dig == 10 else dig
        if dig != int(cpf[tam]):
            return False
    return True


def valida_cnpj(cnpj: str) -> bool:
    cnpj = so_digitos(cnpj)
    if len(cnpj) != 14 or cnpj == cnpj[0] * 14:
        return False
    pesos1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    pesos2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    for pesos, pos in ((pesos1, 12), (pesos2, 13)):
        soma = sum(int(cnpj[i]) * pesos[i] for i in range(pos))
        resto = soma % 11
        dig = 0 if resto < 2 else 11 - resto
        if dig != int(cnpj[pos]):
            return False
    return True


def valida_cpf_cnpj(numero: str) -> bool:
    """True se for CPF (11) ou CNPJ (14) com dígitos verificadores válidos."""
    num = so_digitos(numero)
    if len(num) == 11:
        return valida_cpf(num)
    if len(num) == 14:
        return valida_cnpj(num)
    return False
